libraries/3rdparty/lwip/src/portable/arch/sys_arch.c

```diff
@@ -578,6 +578,46 @@ sys_thread_t xReturn;
     return xReturn;
 }
 
+#if LWIP_NETCONN_SEM_PER_THREAD
+#if configNUM_THREAD_LOCAL_STORAGE_POINTERS > 0
+
+/*---------------------------------------------------------------------------*
+ * Routine:  sys_arch_netconn_sem_get
+ *---------------------------------------------------------------------------*
+ * Description:
+ *      Lookup the task-specific semaphore; create one if necessary.
+ *      The semaphore pointer lives in the 0th slot of the
+ *      TCB local storage array.  Once allocated, it is never released.
+ *---------------------------------------------------------------------------*/
+sys_sem_t *
+sys_arch_netconn_sem_get(void)
+{
+    void* ret;
+    TaskHandle_t task = xTaskGetCurrentTaskHandle();
+    configASSERT( task != NULL );
+
+    ret = pvTaskGetThreadLocalStoragePointer( task, 0 );
+    if( ret == NULL )
+    {
+        sys_sem_t *sem;
+        err_t err;
+        /* allocate memory for this semaphore */
+        sem = mem_malloc( sizeof( sys_sem_t ) );
+        configASSERT( sem != NULL );
+        err = sys_sem_new( sem, 0 );
+        configASSERT( err == ERR_OK );
+        configASSERT( sys_sem_valid( sem ) );
+        vTaskSetThreadLocalStoragePointer( task, 0, sem );
+        ret = sem;
```

Comment:
Is there any chance that 2 tasks both try to grab the semaphore before it's been created and cause this semaphore to be double allocated?

Reply:
The semaphore is per-task, so if two tasks grab semaphores, they get two different semaphores.  The lwIP approach to concurrency is non-intuitive.